transports/webrtc/src/tokio/udp_mux.rs

```diff
@@ -0,0 +1,575 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use async_trait::async_trait;
+use futures::{
+    channel::oneshot,
+    future::{BoxFuture, FutureExt, OptionFuture},
+    stream::FuturesUnordered,
+    StreamExt,
+};
+use stun::{
+    attributes::ATTR_USERNAME,
+    message::{is_message as is_stun_message, Message as STUNMessage},
+};
+use thiserror::Error;
+use tokio::{io::ReadBuf, net::UdpSocket};
+use webrtc::ice::udp_mux::{UDPMux, UDPMuxConn, UDPMuxConnParams, UDPMuxWriter};
+use webrtc::util::{Conn, Error};
+
+use std::{
+    collections::{HashMap, HashSet},
+    io,
+    io::ErrorKind,
+    net::SocketAddr,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use crate::tokio::req_res_chan;
+
+const RECEIVE_MTU: usize = 8192;
+
+/// A previously unseen address of a remote who've sent us an ICE binding request.
+#[derive(Debug)]
+pub struct NewAddr {
+    pub addr: SocketAddr,
+    pub ufrag: String,
+}
+
+/// An event emitted by [`UDPMuxNewAddr`] when it's polled.
+#[derive(Debug)]
+pub enum UDPMuxEvent {
+    /// Connection error. UDP mux should be stopped.
+    Error(std::io::Error),
+    /// Got a [`NewAddr`] from the socket.
+    NewAddr(NewAddr),
+}
+
+/// A modified version of [`webrtc::ice::udp_mux::UDPMuxDefault`].
+///
+/// - It has been rewritten to work without locks and channels instead.
+/// - It reports previously unseen addresses instead of ignoring them.
+pub struct UDPMuxNewAddr {
+    udp_sock: UdpSocket,
+
+    listen_addr: SocketAddr,
+
+    /// Maps from ufrag to the underlying connection.
+    conns: HashMap<String, UDPMuxConn>,
+
+    /// Maps from socket address to the underlying connection.
+    address_map: HashMap<SocketAddr, UDPMuxConn>,
+
+    /// Set of the new addresses to avoid sending the same address multiple times.
+    new_addrs: HashSet<SocketAddr>,
+
+    /// `true` when UDP mux is closed.
+    is_closed: bool,
+
+    send_buffer: Option<(Vec<u8>, SocketAddr, oneshot::Sender<Result<usize, Error>>)>,
+
+    close_futures: FuturesUnordered<BoxFuture<'static, ()>>,
+    write_future: OptionFuture<BoxFuture<'static, ()>>,
+
+    close_command: req_res_chan::Receiver<(), Result<(), Error>>,
+    get_conn_command: req_res_chan::Receiver<String, Result<Arc<dyn Conn + Send + Sync>, Error>>,
+    remove_conn_command: req_res_chan::Receiver<String, ()>,
+    registration_command: req_res_chan::Receiver<(UDPMuxConn, SocketAddr), ()>,
+    send_command: req_res_chan::Receiver<(Vec<u8>, SocketAddr), Result<usize, Error>>,
+
+    udp_mux_handle: Arc<UdpMuxHandle>,
+    udp_mux_writer_handle: Arc<UdpMuxWriterHandle>,
+}
+
+impl UDPMuxNewAddr {
+    pub fn listen_on(addr: SocketAddr) -> Result<Self, io::Error> {
+        // XXX: `UdpSocket::bind` is async, so use a std socket and convert
+        let std_sock = std::net::UdpSocket::bind(addr)?;
```

Comment:
Would it make sense to call `take_error` after binding here?

https://docs.rs/tokio/latest/tokio/net/struct.UdpSocket.html#method.take_error

https://github.com/libp2p/rust-libp2p/blob/4d4833f8c2622018b5b775896905a534b8624d3c/transports/tcp/src/provider/tokio.rs#L58-L60

Reply:
what for?

```
If so_error is nonzero when the process calls read and there is no data to return, read returns–1 with errno set to the value of so_error (p. 516 of TCPv2). The value of so_error is then reset to 0. If there is data queued for the socket, that data is returned by read instead of the error condition. If so_error is nonzero when the process calls write, –1 is returned with errno set to the value of so_error (p. 495 of TCPv2) and so_error is reset to 0.
```

indicates that it's set once there's a failed read OR write op.